### ISSUE TITLE
[Proposal] Add type definitions via d.ts files around public API surface

### ISSUE THREAD

**Issue by emilio-martinez** [open-wc/open-wc#572]
### Things I've reviewed

- Read the `types.md`
- Read through some previous issues
- Read through some PRs such as https://github.com/open-wc/open-wc/pull/277

### Preamble

In my opinion, the current required setup to get type hinting is way too cumbersome. Perhaps less so for Javascript, for sure; however, having to go as far as changing my build process to get type hinting _only for this project's packages_ is kind of a no-go. I'm not going to propose changing this project to Typescript—that's kind of a separate matter, but I do believe getting a more frictionless experience with types is needed.

### Proposal

I'd like to propose exposing type definition files (`*.d.ts`) that cover the public API surface of this project. This is similar to what DefinitelyTyped does, but without having to install a separate package. As an example, several of [sindresorhus](https://github.com/sindresorhus)'s repos follow this pattern.

**Benefit**
- Types work out-of-the-box without having to convert the entire source of the project.
- Can be done gradually and without impacting current source.
- `// @ts-ignore` comments in Javascript code would no longer be required.
- Only needs to cover public API.

**Downsides**
- Types could fall out of sync because not tied to the source, but on the bright side, this only applied to the public API.

**Comment by AdamVig** [comment]
See also #538, which has the same goal as this issue, but no proposed implementation.

**Comment by LarsDenBakker** [comment]
Thanks for your proposal. To make sure I understand this correctly: our jsdocs work when importing from JS, but not when importing from TS?

**Comment by emilio-martinez** [comment]
Essentially, yes. As the `types.md` document says, there's a way to configure the types to come through to typescript, but it requires a change in build process that's only required for this project's packages, which in many circumstances is a no-go.

Type definitions files (`d.ts`) are picked up by editors such as VSCode for Javascript and Typescript equally, with the pros and cons outlined in the description of this issue.

**Comment by daKmoR** [comment]
yeah I fully agree that it's not as simple as it should be 😭 
such a pity that support is not better for it... 

> Types could fall out of sync because not tied to the source

yeah and that goes faster then you think 🙈 

is there a way to auto-generate definition files from JsDoc?
I mean we run `tsc` to make sure all our types are correct... it's just that we can't easily write them to definition files... I know that webpack does it like that... but their setup seems rather complicated... anyone volunteers to write such a tool? 😬 

@runem you are using the typescript lang server for the analyzer right? do you see a chance that typescript or the lang server somehow output definition files? (for code that has correct types but in JSDoc)

**Comment by bennypowers** [comment]
have you added a jsconfig.json to your consuming repo?

**Comment by LarsDenBakker** [comment]
Emitting .d.ts files from jsdoc would be great indeed.

**Comment by emilio-martinez** [comment]
Yeah, I believe there's been a feature request open for a bit wanting .d.ts output from JS+JSDoc when "declarations" is true in a tsconfig, but it hasn't landed yet, unfortunately. (Sorry, I would provide an issue number but I'm on my phone)

I may have time to fiddle around with the idea later this week; I'll report back any findings.

**Comment by benjamind** [comment]
@daKmoR I've had some success generating tsd from the jsdoc in the open-wc project using this https://www.npmjs.com/package/tsd-jsdoc

It required a little manual cleanup after generation but it very quickly got me to a working set of typescript declarations for the parts of openwc I'm using.

**Comment by daKmoR** [comment]
soooo @runem made a fantastic script that might get as exactly where we want... until tsc supports it directly 💪 

see it here... pls help verify if the output is what you would expect/need
https://github.com/open-wc/open-wc/pull/678

**Comment by AdamVig** [comment]
Seems fine from an end-user perspective. Having automatically-generated types is a huge win.

From a maintainer perspective, it seems like it would be easier all around to do a gradual migration to TypeScript, but I assume that the maintainers here have good reasons to keep using JavaScript.

**Comment by benjamind** [comment]
This is pretty awesome @runem. I took a quick look at the output and tried it use it in a project, while it does stop TS from complaining we are basically losing all typesafety on imported dependencies.

This might be due to the way ts handles the inline `import('lit-html').TemplateResult` statements in the JSDoc, or possibly its just not got the correct module resolution when running the compilation. I've had a little play around with it to see if I could figure out which thing it is, but didn't get very far before I ran out of time.

I also tried modifying the import statements to use the `@typedef` syntax to import the lit-html types and map them to local typedefs which are then returned by the jsdoc but that didn't work either.

Will try and find more time to help out on this when I can.
